win32ss/user/user32/misc/dllmain.c

```diff
@@ -544,18 +557,19 @@ DllMain(
 
             if (!gfServerProcess)
             {
-#if WIN32K_ISNT_BROKEN
-               InitializeImmEntryTable();
-#else
-               /* imm32 takes a refcount and prevents us from unloading */
-               LoadLibraryW(L"user32");
-#endif
-               //
-               // Wine is stub and throws an exception so save this for real Imm32.dll testing!!!!
-               //
-               //gImmApiEntries.pImmRegisterClient(&gSharedInfo, ghImm32);
+                HINSTANCE hImm32 = NULL;
+
+                if (gpsi && (gpsi->dwSRVIFlags & SRVINFO_IMM32))
+                {
+                    WCHAR szImmFile[MAX_PATH];
+                    InitializeImmEntryTable();
+                    GetImmFileName(szImmFile, _countof(szImmFile));
+                    hImm32 = GetModuleHandleW(szImmFile);
+                }
+
+                if (!IMM_FN(ImmRegisterClient)(&gSharedInfo, hImm32))
```

Comment:
`gSharedInfo` is `user32`'s global variable.
Can we pass it to `imm32`'s `ImmRegisterClient`?

Reply:
If imm32 needs to access it, sure.

Reply:
But access violation happens.

Reply:
Access violation is fixed.